sql/core/src/main/scala/org/apache/spark/sql/execution/SkewShuffleRowRDD.scala

```diff
@@ -0,0 +1,147 @@
+
+class SkewCoalescedPartitioner(
+        val parent: Partitioner,
+        val partitionStartIndices: Array[(Int, Int)])
+  extends Partitioner {
+
+  @transient private lazy val parentPartitionMapping: Array[Int] = {
+    val n = parent.numPartitions
+    val result = new Array[Int](n)
+    for (i <- 0 until partitionStartIndices.length) {
+      val start = partitionStartIndices(i)._2
+      val end = if (i < partitionStartIndices.length - 1) partitionStartIndices(i + 1)._2 else n
+      for (j <- start until end) {
+        result(j) = i
+      }
+    }
+    result
+  }
+
+  override def numPartitions: Int = partitionStartIndices.length
+
+  override def getPartition(key: Any): Int = {
+    parentPartitionMapping(parent.getPartition(key))
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case c: SkewCoalescedPartitioner =>
+      c.parent == parent &&
+        c.partitionStartIndices.zip(partitionStartIndices).
+          forall( r => r match {
+            case (x, y) => (x._1 == y._1 && x._2 == y._2)
+            })
+    case _ =>
+      false
+  }
+
+  override def hashCode(): Int = 31 * parent.hashCode() + partitionStartIndices.hashCode()
+}
+
+ /**
+  * if mapIndex is -1, same as ShuffledRowRDDPartition
+  * if mapIndex > -1 ,only read one block of mappers.
+  */
+private final class SkewShuffledRowRDDPartition(
+    val postShufflePartitionIndex: Int,
+    val mapIndex: Int,
+    val startPreShufflePartitionIndex: Int,
+    val endPreShufflePartitionIndex: Int) extends Partition {
+  override val index: Int = postShufflePartitionIndex
+
+  override def hashCode(): Int = postShufflePartitionIndex
+
+  override def equals(other: Any): Boolean = super.equals(other)
+}
+
+ /**
+  * only use for skew data join. In join case , need fetch the same partition of
+  * left output and rigth output together. but when some partiton have bigger data than
+  * other partitions, it occur data skew . in the case , we need a specialized RDD to handling this.
+  * in skew partition side,we don't produce one partition, because one partition produce
+  * one task deal so much data is too slaw . but produce per-stage mapping task num parititons.
+  * one task only deal one mapper data. in other no skew side. In order to deal with the
+  * corresponding skew partition , we need produce same partition per-stage parititon num
+  * times.(Equivalent to broadcoast this partition)
+  *
+  * other no skew partition, then deal like ShuffledRowRDD
+  */
+class SkewShuffleRowRDD(
+    var dependency1: ShuffleDependency[Int, InternalRow, InternalRow],
+    partitionStartIndices: Array[(Int, Int, Int)])
+  extends ShuffledRowRDD ( dependency1, None) {
+
+  private[this] val numPreShufflePartitions = dependency.partitioner.numPartitions
+
+  override def getPartitions: Array[Partition] = {
+    val partitions = ArrayBuffer[Partition]()
+    var partitionIndex = -1
+    for(i <- 0 until partitionStartIndices.length ) {
```

Comment:
` for(i <- 0 until partitionStartIndices.length )` -> `  for (i <- partitionStartIndices.indices) `